service/daemon.go

```diff
@@ -0,0 +1,187 @@
+	"github.com/hashicorp/consul/api"
+	"github.com/xiaoyao1991/chukonu/core"
+	"golang.org/x/net/context"
+)
+
+const ChukonuImage = "chukonu"
+const BytesInKB = 1024
+const BytesInMB = 1024 * 1024
+const BytesInGB = 1024 * 1024 * 1024
+const numAcceptors = 100
+
+type Daemon struct {
+	hostname        string
+	consulAddr      string
+	consul          *api.Client
+	cadvisorBaseUrl string
+	cadvisor        *cadvisor.Client
+	docker          *docker.Client
+}
+
+func NewDaemon(cadvisorBaseUrl string, consulAddr string) Daemon {
+	consul, err := api.NewClient(&api.Config{Address: consulAddr})
+	if err != nil {
+		panic(err)
+	}
+	cadvisor, err := cadvisor.NewClient(cadvisorBaseUrl)
+	if err != nil {
+		panic(err)
+	}
+	docker, err := docker.NewEnvClient()
+	if err != nil {
+		panic(err)
+	}
+
+	hostname, _ := os.Hostname()
+
+	return Daemon{
+		hostname:        hostname,
+		consulAddr:      consulAddr,
+		consul:          consul,
+		cadvisorBaseUrl: cadvisorBaseUrl,
+		cadvisor:        cadvisor,
+		docker:          docker,
+	}
+}
+
+func (d Daemon) SetupTestPlan(config core.ChukonuConfig) error {
+	kv := d.consul.KV()
+
+	concurrencyB := make([]byte, 8)
+	iterationsB := make([]byte, 8)
+	totalTimeoutB := make([]byte, 8)
+
+	binary.LittleEndian.PutUint64(concurrencyB, uint64(config.Concurrency))
+	binary.LittleEndian.PutUint64(iterationsB, uint64(config.Iterations))
+	binary.LittleEndian.PutUint64(totalTimeoutB, uint64(config.TotalTimeout.Nanoseconds()))
+
+	kvpair := &api.KVPair{
+		Key:   fmt.Sprintf("%s/config/concurrency", config.TenantId),
+		Value: concurrencyB,
+	}
+	_, err := kv.Put(kvpair, nil)
+	if err != nil {
+		return err
+	}
+
+	kvpair = &api.KVPair{
+		Key:   fmt.Sprintf("%s/config/iterations", config.TenantId),
+		Value: iterationsB,
+	}
+	_, err = kv.Put(kvpair, nil)
+	if err != nil {
+		return err
+	}
+
+	kvpair = &api.KVPair{
+		Key:   fmt.Sprintf("%s/config/timeout", config.TenantId),
+		Value: totalTimeoutB,
+	}
+	_, err = kv.Put(kvpair, nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d Daemon) SpawnNewContainer(tenantId string) string {
+	ctx := context.Background()
+
+	// TODO: lock consul KV table
+
+	resp, err := d.docker.ContainerCreate(ctx, &container.Config{
+		Image: ChukonuImage,
+		Cmd:   []string{"-tenant", tenantId, "-cadvisor", d.cadvisorBaseUrl, "-consul", d.consulAddr},
+	}, &container.HostConfig{
+		Resources: container.Resources{
+			// TODO: how do we decide what limit? should we smart calculate?
+			Memory:     int64(16 * BytesInMB),
+			MemorySwap: int64(20 * BytesInMB),
+			CPUQuota:   int64(50000),
+			CPUPeriod:  int64(100000),
+		},
+	}, nil, "")
+	if err != nil {
+		panic(err)
+	}
+
+	if err := d.docker.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{}); err != nil {
+		panic(err)
+	}
+
+	// TODO: unlock consul KV table? or should we unlock it in lifecycle
+
+	return resp.ID
+}
+
+func (d Daemon) ShouldSpawnNewContainer(tenantId string) bool {
```

Comment:
I wonder would there be a race condition since we update the `workercount` in lifecycle and checked here. Chances are you check and spawn container too fast in daemon and spawn too many containers before each one update their portion of `workercount` in lifecycle. Am I understanding this correctly?